File: backend/src/common/production-stages.ts
/**
 * Отметка производства: по ИЗДЕЛИЯМ, а не по видам работ (26.08.2026).
 *
 * Было пять шагов, потом три вида работ — но цеху и это лишнее: мастер
 * должен показать, что конкретное изделие изготовлено, а не расписывать,
 * какую операцию он делал (запрос пользователя: «уберём этапы работ,
 * просто покажем, что сделано по конкретной готовой продукции»).
 *
 * Виды работ (CUTTING/ASSEMBLY/PAINTING) НЕ исчезли из системы — они
 * остались там, где действительно нужны: нормы труда, ставка часа по
 * каждому виду работ, подряд. Себестоимость от отметок цеха не зависит
 * вообще (проверено: order-costing.service.ts не читает ProductionStage),
 * поэтому убрать их из отметки безопасно.
 *
 * Сырьё и ТМЦ (Article.isMaterialResale) в производство не попадают:
 * завод их не изготавливает, а перепродаёт. В активных заказах это
 * 378 позиций из 1942 — пятая часть очереди цеха была мусором.
 */

export type OrderStageCodeValue = 'PRODUCTION';
export type RoutingStageValue = 'CUTTING' | 'ASSEMBLY' | 'PAINTING';

export interface StageStep {
  code: OrderStageCodeValue;
  routingStage: RoutingStageValue;
  key: string;
  label: string;
}

/**
 * Виды работ — справочник для НОРМ и СТАВОК, не для отметки цеха.
 * Мастер их не выбирает; они нужны калькуляции (норма чел×часы на вид
 * работ) и подряду (какую работу отдали на сторону).
 */
export const STAGE_STEPS: StageStep[] = [
  { code: 'PRODUCTION', routingStage: 'CUTTING', key: 'PRODUCTION:CUTTING', label: 'Резка' },
  { code: 'PRODUCTION', routingStage: 'ASSEMBLY', key: 'PRODUCTION:ASSEMBLY', label: 'Сборка / сварка / обшивка' },
  { code: 'PRODUCTION', routingStage: 'PAINTING', key: 'PRODUCTION:PAINTING', label: 'Зачистка / покраска' },
];

export const ORDER_STAGE_CODES: OrderStageCodeValue[] = ['PRODUCTION'];
export const ROUTING_STAGES: RoutingStageValue[] = ['CUTTING', 'ASSEMBLY', 'PAINTING'];

/** Ключ шага: «PRODUCTION:CUTTING» — по нему сопоставляются записи БД */
export function stepKey(code: string, routingStage?: string | null): string {
  return routingStage ? `${code}:${routingStage}` : code;
}

/**
 * Отметка изделия: этап всегда PRODUCTION, вид работ НЕ обязателен —
 * мастер отмечает «изделие изготовлено» целиком. Вид работ принимается,
 * если его всё же прислали (совместимость и возможные частичные отметки).
 */
export function stageShapeError(code: string, routingStage?: string | null): string | null {
  if (!ORDER_STAGE_CODES.includes(code as OrderStageCodeValue)) {
    return `Неизвестный этап: ${code}. Допустимо: ${ORDER_STAGE_CODES.join(', ')}`;
  }
  if (routingStage && !ROUTING_STAGES.includes(routingStage as RoutingStageValue)) {
    return `Неизвестный вид работ: ${routingStage}. Допустимо: ${ROUTING_STAGES.join(', ')}`;
  }
  return null;
}

/**
 * Готовность заказа = сколько его ИЗДЕЛИЙ изготовлено (26.08.2026).
 *
 * Мерой служит список изделий заказа, а не отметки в базе: строки этапов
 * создаются лениво, и `every(done)` по ним пропускал бы заказ с одной
 * отметкой вперёд, минуя остальные изделия. Сырьё и ТМЦ в список не
 * входят — их не изготавливают.
 */
export function stageProgress(
  rows: Array<{ orderLineId?: string | null; status: string }>,
  /** Позиции-изделия заказа (без сырья и ТМЦ). Пусто — считать нечего */
  productLineIds: string[] = [],
): { allDone: boolean; anyStarted: boolean; doneCount: number; totalSteps: number } {
  const doneLines = new Set<string>();
  let anyStarted = false;
  for (const r of rows) {
    const st = r.status.toLowerCase();
    if (st === 'done' || st === 'in_progress') anyStarted = true;
    if (st === 'done' && r.orderLineId) doneLines.add(r.orderLineId);
  }

  const needed = productLineIds.length;
  const doneCount = productLineIds.filter((id) => doneLines.has(id)).length;
  return {
    // Заказ без изделий (только сырьё) готовым по цеху не становится
    allDone: needed > 0 && doneCount === needed,
    anyStarted,
    doneCount,
    totalSteps: needed,
  };
}

export const DEFAULT_STAGE_TRACKING_THRESHOLD = 5;

export function resolveTrackingMode(lineCount: number, threshold: number): 'ORDER' | 'LINE' {
  return lineCount > threshold ? 'LINE' : 'ORDER';
}

export interface LineNorm {
  orderLineId: string;
  /** Нормативная трудоёмкость позиции на переделе: чел × часы × количество */
  normManHours: number;
}

export interface HoursAllocation {
  orderLineId: string;
  sharePct: number;
  hours: number;
}

/**
 * Раскладка фактических часов по позициям пропорционально нормам (§2.3).
 * Мастер вводит один факт на заказ — себестоимость всё равно получает разбивку
 * по изделиям, без второго ввода.
 *
 * Если норм нет вовсе (новое изделие без нормирования), часы делятся поровну —
 * это честнее, чем отдать всё первой позиции или потерять их совсем.
 */
export function allocateActualHours(actualHours: number, lines: LineNorm[]): HoursAllocation[] {
  if (lines.length === 0) return [];
  const totalNorm = lines.reduce((s, l) => s + Math.max(0, l.normManHours), 0);
  const weights = totalNorm > 0
    ? lines.map((l) => Math.max(0, l.normManHours) / totalNorm)
    : lines.map(() => 1 / lines.length);

  const round3 = (n: number) => Math.round(n * 1000) / 1000;
  const allocated = lines.map((l, i) => ({
    orderLineId: l.orderLineId,
    sharePct: Math.round(weights[i] * 10000) / 100,
    hours: round3(actualHours * weights[i]),
  }));

  // Хвост округления кладём в самую крупную позицию: сумма долей должна
  // сходиться с введённым фактом до копейки, иначе план/факт не сойдётся
  const diff = round3(actualHours - allocated.reduce((s, a) => s + a.hours, 0));
  if (diff !== 0) {
    const biggest = allocated.reduce((a, b) => (b.hours > a.hours ? b : a), allocated[0]);
    biggest.hours = round3(biggest.hours + diff);
  }
  return allocated;
}

File: backend-go/internal/orderstate/production_stages.go
package orderstate

import "math"

// orderStageCodes — ТОЛЬКО PRODUCTION: DESIGN/SUPPLY существуют как
// значения enum OrderStageCode в БД (легаси), но отметка цеха по вехам
// «убрана» решением 26.08.2026 — stageShapeError в оригинале принимает
// исключительно 'PRODUCTION', не все три значения enum.
var orderStageCodes = []string{"PRODUCTION"}
var routingStages = []string{"CUTTING", "ASSEMBLY", "PAINTING"}

func contains(list []string, v string) bool {
	for _, x := range list {
		if x == v {
			return true
		}
	}
	return false
}

// StageShapeError — код этапа всегда PRODUCTION/DESIGN/SUPPLY (API-код, уже
// переведённый из русской метки БД), вид работ CUTTING/ASSEMBLY/PAINTING
// необязателен. Возвращает "" (нет ошибки) или готовый текст сообщения.
func StageShapeError(code string, routingStage *string) string {
	if !contains(orderStageCodes, code) {
		return "Неизвестный этап: " + code + ". Допустимо: " + joinComma(orderStageCodes)
	}
	if routingStage != nil && *routingStage != "" && !contains(routingStages, *routingStage) {
		return "Неизвестный вид работ: " + *routingStage + ". Допустимо: " + joinComma(routingStages)
	}
	return ""
}

func joinComma(list []string) string {
	out := ""
	for i, s := range list {
		if i > 0 {
			out += ", "
		}
		out += s
	}
	return out
}

// StageRow — запись этапа для расчёта готовности (status уже в нижнем
// регистре API-значения: not_started/in_progress/done — как отдаёт
// StageStatusDBToAPI-обратное сравнение делает вызывающий код).
type StageRow struct {
	StageCode    string
	RoutingStage *string
	OrderLineID  *string
	Status       string // lowercase: "done" | "in_progress" | "not_started"
}

type StageProgressResult struct {
	AllDone    bool
	AnyStarted bool
	DoneCount  int
	TotalSteps int
}

// StageProgress — готовность заказа = сколько его ИЗДЕЛИЙ изготовлено.
// Мерой служит список изделий заказа (productLineIds), а не отметки в
// базе — они создаются лениво.
func StageProgress(rows []StageRow, productLineIDs []string) StageProgressResult {
	doneLines := map[string]bool{}
	anyStarted := false
	for _, r := range rows {
		st := r.Status
		if st == "done" || st == "in_progress" {
			anyStarted = true
		}
		if st == "done" && r.OrderLineID != nil {
			doneLines[*r.OrderLineID] = true
		}
	}
	needed := len(productLineIDs)
	doneCount := 0
	for _, id := range productLineIDs {
		if doneLines[id] {
			doneCount++
		}
	}
	return StageProgressResult{
		AllDone:    needed > 0 && doneCount == needed,
		AnyStarted: anyStarted,
		DoneCount:  doneCount,
		TotalSteps: needed,
	}
}

const DefaultStageTrackingThreshold = 5

func ResolveTrackingMode(lineCount, threshold int) string {
	if lineCount > threshold {
		return "LINE"
	}
	return "ORDER"
}

type LineNorm struct {
	OrderLineID  string
	NormManHours float64
}

type HoursAllocation struct {
	OrderLineID string  `json:"orderLineId"`
	SharePct    float64 `json:"sharePct"`
	Hours       float64 `json:"hours"`
}

func round3s(n float64) float64 { return math.Round(n*1000) / 1000 }

// AllocateActualHours — раскладка фактических часов по позициям
// пропорционально нормам; нет норм — делим поровну. Хвост округления
// уходит в самую крупную позицию, чтобы сумма долей сходилась до копейки.
func AllocateActualHours(actualHours float64, lines []LineNorm) []HoursAllocation {
	if len(lines) == 0 {
		return []HoursAllocation{}
	}
	totalNorm := 0.0
	for _, l := range lines {
		if l.NormManHours > 0 {
			totalNorm += l.NormManHours
		}
	}
	weights := make([]float64, len(lines))
	for i, l := range lines {
		if totalNorm > 0 {
			w := l.NormManHours
			if w < 0 {
				w = 0
			}
			weights[i] = w / totalNorm
		} else {
			weights[i] = 1.0 / float64(len(lines))
		}
	}

	allocated := make([]HoursAllocation, len(lines))
	sum := 0.0
	for i, l := range lines {
		hours := round3s(actualHours * weights[i])
		allocated[i] = HoursAllocation{
			OrderLineID: l.OrderLineID,
			SharePct:    math.Round(weights[i]*10000) / 100,
			Hours:       hours,
		}
		sum += hours
	}

	diff := round3s(actualHours - sum)
	if diff != 0 {
		biggest := 0
		for i := 1; i < len(allocated); i++ {
			if allocated[i].Hours > allocated[biggest].Hours {
				biggest = i
			}
		}
		allocated[biggest].Hours = round3s(allocated[biggest].Hours + diff)
	}
	return allocated
}
